runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/work/WorkProcessingContext.java

```diff
@@ -0,0 +1,129 @@
+import org.apache.beam.sdk.annotations.Internal;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.base.Preconditions;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.joda.time.Instant;
+
+/** Context to process {@link WorkItem} */
+@AutoValue
+@Internal
+public abstract class WorkProcessingContext {
+  public static WorkProcessingContext.Builder builder() {
+    return new AutoValue_WorkProcessingContext.Builder();
+  }
+
+  public static WorkProcessingContext.Builder builder(
+      String computationId,
+      BiFunction<String, KeyedGetDataRequest, Optional<KeyedGetDataResponse>> keyedDataFetcher) {
+    return builder()
+        .setComputationId(computationId)
+        .setKeyedDataFetcher(request -> keyedDataFetcher.apply(computationId, request));
+  }
+
+  public static WorkProcessingContext.Builder builder(
+      String computationId, GetDataStream getDataStream) {
+    return builder()
+        .setComputationId(computationId)
+        .setGetDataStream(getDataStream)
+        .setKeyedDataFetcher(
+            request -> Optional.ofNullable(getDataStream.requestKeyedData(computationId, request)));
+  }
+
+  public abstract String computationId();
+
+  public abstract Instant inputDataWatermark();
+
+  public abstract @Nullable Instant synchronizedProcessingTime();
+
+  public abstract @Nullable Instant outputDataWatermark();
+
+  public abstract @Nullable GetDataStream getDataStream();
+
+  /** {@link WorkItem} being processed. */
+  public abstract WorkItem workItem();
+
+  /**
+   * {@link GetDataStream} that connects to the backend Windmill worker handling the {@link
+   * WorkItem}.
+   */
+  public abstract Function<KeyedGetDataRequest, Optional<KeyedGetDataResponse>> keyedDataFetcher();
+
+  /**
+   * {@link WorkCommitter} that commits completed work to the backend Windmill worker handling the
+   * {@link WorkItem}.
+   */
+  public abstract Consumer<Commit> workCommitter();
+
+  public final void queueCommit(Commit commit) {
+    workCommitter().accept(commit);
+  }
+
+  @Memoized
+  public ShardedKey shardedKey() {
+    return ShardedKey.create(workItem().getKey(), workItem().getShardingKey());
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+    public abstract Builder setComputationId(String value);
+
+    public abstract Builder setInputDataWatermark(Instant value);
+
+    public abstract Builder setSynchronizedProcessingTime(@Nullable Instant value);
+
+    public abstract Builder setOutputDataWatermark(@Nullable Instant value);
+
+    public abstract Builder setWorkItem(WorkItem value);
+
+    public abstract Builder setKeyedDataFetcher(
+        Function<KeyedGetDataRequest, Optional<KeyedGetDataResponse>> value);
+
+    public abstract Builder setWorkCommitter(Consumer<Commit> value);
+
+    public abstract Builder setGetDataStream(GetDataStream value);
+
+    abstract WorkProcessingContext autoBuild();
```

Comment:
private?

Reply:
autoBuild needs to be abstract since it is handled by autovalue codegen

Reply:
will try to use protected and see if it works
previously was following this guide https://github.com/google/auto/blob/main/value/userguide/builders-howto.md#-validate-property-values